File: src/core/output.ts
import { stringify as stringifyYaml } from "yaml";

export const CLI_API_VERSION = "agent-identity.cli/v1";
export type OutputFormat = "json" | "yaml" | "table";

export interface SuccessEnvelope<T = unknown> {
  api_version: typeof CLI_API_VERSION;
  kind: string;
  data: T;
  request_id?: string;
  warnings: string[];
}

export interface FailureEnvelope {
  api_version: typeof CLI_API_VERSION;
  error: {
    code: string;
    message: string;
    remediation?: Record<string, unknown>;
  };
  request_id?: string;
}

export function parseOutputFormat(value: string): OutputFormat {
  const normalized = value.trim().toLowerCase();
  if (normalized === "json" || normalized === "yaml" || normalized === "table") {
    return normalized;
  }
  throw new Error("output must be table, json, or yaml");
}

export function serializeSuccess(
  format: OutputFormat,
  kind: string,
  data: unknown,
  requestId = "",
  warnings: string[] = []
): string {
  const envelope: SuccessEnvelope = {
    api_version: CLI_API_VERSION,
    kind,
    data,
    warnings
  };
  if (requestId !== "") {
    envelope.request_id = requestId;
  }
  if (format === "yaml") {
    return stringifyYaml(envelope);
  }
  if (format === "table") {
    const rows = [`KIND\t${kind}`];
    if (requestId !== "") {
      rows.push(`REQUEST_ID\t${requestId}`);
    }
    writeTableRows(rows, "DATA", data);
    for (const warning of warnings) {
      rows.push(`WARNING\t${warning}`);
    }
    return `${rows.join("\n")}\n`;
  }
  return `${JSON.stringify(envelope)}\n`;
}

export function serializeFailure(
  code: string,
  message: string,
  requestId = "",
  remediation?: Record<string, unknown>
): string {
  const error: FailureEnvelope["error"] = { code, message };
  if (remediation !== undefined) {
    error.remediation = remediation;
  }
  const envelope: FailureEnvelope = { api_version: CLI_API_VERSION, error };
  if (requestId !== "") {
    envelope.request_id = requestId;
  }
  return `${JSON.stringify(envelope)}\n`;
}

function writeTableRows(rows: string[], prefix: string, value: unknown): void {
  if (isRecord(value)) {
    for (const key of Object.keys(value).sort()) {
      writeTableRows(rows, key, value[key]);
    }
    return;
  }
  const encoded = JSON.stringify(value);
  rows.push(`${prefix.toUpperCase()}\t${stripJsonStringQuotes(encoded)}`);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function stripJsonStringQuotes(value: string | undefined): string {
  if (value === undefined) {
    return "null";
  }
  return value.startsWith('"') && value.endsWith('"') ? value.slice(1, -1) : value;
}

File: internal/cli/output/output.go
package output

import (
	"encoding/json"
	"fmt"
	"io"
	"sort"
	"strings"
	"sync"

	"gopkg.in/yaml.v3"
)

const APIVersion = "agent-identity.cli/v1"

var formatState = struct {
	sync.RWMutex
	value string
}{value: "json"}

func SetFormat(value string) error {
	value = strings.ToLower(strings.TrimSpace(value))
	if value != "json" && value != "yaml" && value != "table" {
		return fmt.Errorf("output must be table, json, or yaml")
	}
	formatState.Lock()
	formatState.value = value
	formatState.Unlock()
	return nil
}

func currentFormat() string {
	formatState.RLock()
	defer formatState.RUnlock()
	return formatState.value
}

type Success struct {
	APIVersion string   `json:"api_version" yaml:"api_version"`
	Kind       string   `json:"kind" yaml:"kind"`
	Data       any      `json:"data" yaml:"data"`
	RequestID  string   `json:"request_id,omitempty" yaml:"request_id,omitempty"`
	Warnings   []string `json:"warnings" yaml:"warnings"`
}

type Failure struct {
	APIVersion string `json:"api_version" yaml:"api_version"`
	Error      struct {
		Code        string         `json:"code"`
		Message     string         `json:"message"`
		Remediation map[string]any `json:"remediation,omitempty"`
	} `json:"error" yaml:"error"`
	RequestID string `json:"request_id,omitempty" yaml:"request_id,omitempty"`
}

func WriteSuccess(writer io.Writer, kind string, data any, requestID string) error {
	return WriteSuccessWithWarnings(writer, kind, data, requestID, nil)
}

func WriteSuccessWithWarnings(writer io.Writer, kind string, data any, requestID string, warnings []string) error {
	if warnings == nil {
		warnings = []string{}
	}
	result := Success{APIVersion: APIVersion, Kind: kind, Data: normalize(data), RequestID: requestID, Warnings: warnings}
	switch currentFormat() {
	case "yaml":
		encoded, err := yaml.Marshal(result)
		if err != nil {
			return err
		}
		_, err = writer.Write(encoded)
		return err
	case "table":
		_, err := fmt.Fprintf(writer, "KIND\t%s\n", kind)
		if err != nil {
			return err
		}
		if requestID != "" {
			if _, err := fmt.Fprintf(writer, "REQUEST_ID\t%s\n", requestID); err != nil {
				return err
			}
		}
		writeTable(writer, "DATA", result.Data)
		for _, warning := range result.Warnings {
			if _, err := fmt.Fprintf(writer, "WARNING\t%s\n", warning); err != nil {
				return err
			}
		}
		return nil
	default:
		return json.NewEncoder(writer).Encode(result)
	}
}

func normalize(value any) any {
	if raw, ok := value.(json.RawMessage); ok && json.Valid(raw) {
		var decoded any
		if json.Unmarshal(raw, &decoded) == nil {
			return decoded
		}
	}
	return value
}

func writeTable(writer io.Writer, prefix string, value any) {
	value = normalize(value)
	switch item := value.(type) {
	case map[string]any:
		keys := make([]string, 0, len(item))
		for key := range item {
			keys = append(keys, key)
		}
		sort.Strings(keys)
		for _, key := range keys {
			writeTable(writer, key, item[key])
		}
	default:
		encoded, _ := json.Marshal(item)
		_, _ = fmt.Fprintf(writer, "%s\t%s\n", strings.ToUpper(prefix), strings.Trim(string(encoded), `"`))
	}
}

func WriteFailure(writer io.Writer, code, message, requestID string, remediation map[string]any) {
	result := Failure{APIVersion: APIVersion, RequestID: requestID}
	result.Error.Code, result.Error.Message, result.Error.Remediation = code, message, remediation
	if err := json.NewEncoder(writer).Encode(result); err != nil {
		_, _ = fmt.Fprintf(writer, "agent-identity: %s\n", message)
	}
}
